# Import

In [1]:
from io import StringIO
import pandas as pd
import numpy as np

In [2]:
with open('data/data-2018.json') as f:
   data = json.load(f)

In [3]:
data.keys()

dict_keys(['stadiums', 'tvchannels', 'teams', 'groups', 'knockout'])

In [4]:
stadiums = pd.DataFrame(data["stadiums"])
tvchannels = pd.DataFrame(data["tvchannels"])
teams = pd.DataFrame(data["teams"])
groups = pd.DataFrame(data["groups"])
knockout = pd.DataFrame(data["knockout"])

In [5]:
fifa = pd.read_csv('data/fifa_2022.csv')
fife = pd.read_csv('data/fife_2022.csv')
matches = pd.read_csv('data/matches.csv')
wolrdcup = pd.read_csv('data/worldcupmatches2014.txt', sep=';')

In [6]:
groups = groups.T.reset_index()
knockout = knockout.T.reset_index()

In [7]:
stadiums

id                     name              city        lat        lng  \
0    1         Luzhniki Stadium            Moscow  55.715765  37.551522   
1    2          Otkrytiye Arena            Moscow  55.817765  37.440363   
2    3       Krestovsky Stadium  Saint Petersburg  59.972740  30.221408   
3    4      Kaliningrad Stadium       Kaliningrad  54.698157  20.533859   
4    5              Kazan Arena             Kazan  55.820983  49.160966   
5    6  Nizhny Novgorod Stadium   Nizhny Novgorod  56.337287  43.963251   
6    7             Cosmos Arena            Samara  53.278146  50.238854   
7    8          Volgograd Arena         Volgograd  48.734195  44.548345   
8    9           Mordovia Arena           Saransk  54.181795  45.203851   
9   10             Rostov Arena     Rostov-on-Don  47.209581  39.738424   
10  11    Fisht Olympic Stadium             Sochi  43.402011  39.955709   
11  12          Central Stadium     Yekaterinburg  56.832490  60.573585   

                                                image  
0   https://upload.wikimedia.org/wikipedia/commons...  
1   https://upload.wikimedia.org/wikipedia/commons...  
2   https://upload.wikimedia.org/wikipedia/commons...  
3   https://upload.wikimedia.org/wikipedia/commons...  
4   https://upload.wikimedia.org/wikipedia/commons...  
5   https://upload.wikimedia.org/wikipedia/commons...  
6   http://img.fifa.com/image/upload/t_l4/d0mymt1u...  
7   https://upload.wikimedia.org/wikipedia/commons...  
8   https://upload.wikimedia.org/wikipedia/commons...  
9   https://upload.wikimedia.org/wikipedia/commons...  
10  https://upload.wikimedia.org/wikipedia/commons...  
11  https://upload.wikimedia.org/wikipedia/commons...

In [8]:
tvchannels

id       name                                               icon  \
0    1        DR1  https://upload.wikimedia.org/wikipedia/commons...   
1    2        TV2  https://upload.wikimedia.org/wikipedia/commons...   
2    3     BBC UK  https://upload.wikimedia.org/wikipedia/commons...   
3    4     ITV UK  https://upload.wikimedia.org/wikipedia/en/9/92...   
4    5    ITV4 UK  https://upload.wikimedia.org/wikipedia/en/9/92...   
5    6     Cuatro  https://upload.wikimedia.org/wikipedia/commons...   
6    7  Telecinco  https://upload.wikimedia.org/wikipedia/commons...   
7    8      Globo  https://upload.wikimedia.org/wikipedia/pt/2/22...   
8    9    SportTV  https://upload.wikimedia.org/wikipedia/commons...   
9   10   SportTV2  https://upload.wikimedia.org/wikipedia/commons...   
10  11   SportTV3  https://upload.wikimedia.org/wikipedia/commons...   
11  12   Sony Ten  https://upload.wikimedia.org/wikipedia/en/5/5f...   
12  13        Fox  https://upload.wikimedia.org/wikipedia/commons...   
13  14        FS1  https://upload.wikimedia.org/wikipedia/commons...   
14  15    TransTV  https://upload.wikimedia.org/wikipedia/id/6/62...   
15  16     Trans7  https://upload.wikimedia.org/wikipedia/commons...   
16  17        ARD  https://upload.wikimedia.org/wikipedia/commons...   
17  18        ZDF  https://upload.wikimedia.org/wikipedia/commons...   
18  19        ONE  https://upload.wikimedia.org/wikipedia/commons...   
19  20      TVP 1  https://upload.wikimedia.org/wikipedia/commons...   
20  21      TVP 2  https://upload.wikimedia.org/wikipedia/commons...   
21  22  TVP SPORT  https://upload.wikimedia.org/wikipedia/commons...   

      country iso2     lang  
0     Denmark   dk     [da]  
1     Denmark   dk     [da]  
2          UK   en     [en]  
3          UK   en     [en]  
4          UK   en     [en]  
5       Spain   es     [es]  
6       Spain   es     [es]  
7      Brazil   br  [pt-br]  
8      Brazil   br  [pt-br]  
9      Brazil   br  [pt-br]  
10     Brazil   br  [pt-br]  
11      India   en     [en]  
12        USA   en     [en]  
13        USA   en     [en]  
14  Indonesia   id     [id]  
15  Indonesia   id     [id]  
16    Germany   de     [de]  
17    Germany   de     [de]  
18    Germany   de     [de]  
19     Poland   pl     [pl]  
20     Poland   pl     [pl]  
21     Poland   pl     [pl]

In [9]:
teams

id          name fifaCode    iso2  \
0    1        Russia      RUS      ru   
1    2  Saudi Arabia      KSA      sa   
2    3         Egypt      EGY      eg   
3    4       Uruguay      URU      uy   
4    5      Portugal      POR      pt   
5    6         Spain      ESP      es   
6    7       Morocco      MAR      ma   
7    8          Iran      IRN      ir   
8    9        France      FRA      fr   
9   10     Australia      AUS      au   
10  11          Peru      PER      pe   
11  12       Denmark      DEN      dk   
12  13     Argentina      ARG      ar   
13  14       Iceland      ISL      is   
14  15       Croatia      CRO      hr   
15  16       Nigeria      NGA      ng   
16  17        Brazil      BRA      br   
17  18   Switzerland      SUI      ch   
18  19    Costa Rica      CRC      cr   
19  20        Serbia      SRB      rs   
20  21       Germany      GER      de   
21  22        Mexico      MEX      mx   
22  23        Sweden      SWE      se   
23  24   South Korea      KOR      kr   
24  25       Belgium      BEL      be   
25  26        Panama      PAN      pa   
26  27       Tunisia      TUN      tn   
27  28       England      ENG  gb-eng   
28  29        Poland      POL      pl   
29  30       Senegal      SEN      sn   
30  31      Colombia      COL      co   
31  32         Japan      JPN      jp   

                                                 flag         emoji  \
0   https://upload.wikimedia.org/wikipedia/en/thum...       flag-ru   
1   https://upload.wikimedia.org/wikipedia/commons...       flag-sa   
2   https://upload.wikimedia.org/wikipedia/commons...       flag-eg   
3   https://upload.wikimedia.org/wikipedia/commons...       flag-uy   
4   https://upload.wikimedia.org/wikipedia/commons...       flag-pt   
5   https://upload.wikimedia.org/wikipedia/en/thum...       flag-es   
6   https://upload.wikimedia.org/wikipedia/commons...       flag-ma   
7   https://upload.wikimedia.org/wikipedia/commons...       flag-ir   
8   https://upload.wikimedia.org/wikipedia/en/thum...       flag-fr   
9   https://upload.wikimedia.org/wikipedia/commons...       flag-au   
10  https://upload.wikimedia.org/wikipedia/commons...       flag-pe   
11  https://upload.wikimedia.org/wikipedia/commons...       flag-dk   
12  https://upload.wikimedia.org/wikipedia/commons...       flag-ar   
13  https://upload.wikimedia.org/wikipedia/commons...       flag-is   
14  https://upload.wikimedia.org/wikipedia/commons...       flag-hr   
15  https://upload.wikimedia.org/wikipedia/commons...       flag-ng   
16  https://upload.wikimedia.org/wikipedia/en/thum...       flag-br   
17  https://upload.wikimedia.org/wikipedia/commons...       flag-ch   
18  https://upload.wikimedia.org/wikipedia/commons...       flag-cr   
19  https://upload.wikimedia.org/wikipedia/commons...       flag-rs   
20  https://upload.wikimedia.org/wikipedia/en/thum...       flag-de   
21  https://upload.wikimedia.org/wikipedia/commons...       flag-mx   
22  https://upload.wikimedia.org/wikipedia/en/thum...       flag-se   
23  https://upload.wikimedia.org/wikipedia/commons...       flag-kr   
24  https://upload.wikimedia.org/wikipedia/commons...       flag-be   
25  https://upload.wikimedia.org/wikipedia/commons...       flag-pa   
26  https://upload.wikimedia.org/wikipedia/commons...       flag-tn   
27  https://upload.wikimedia.org/wikipedia/en/thum...  flag-england   
28  https://upload.wikimedia.org/wikipedia/en/thum...       flag-pl   
29  https://upload.wikimedia.org/wikipedia/commons...       flag-sn   
30  https://upload.wikimedia.org/wikipedia/commons...       flag-co   
31  https://upload.wikimedia.org/wikipedia/en/thum...       flag-jp   

   emojiString  
0           🇷🇺  
1           🇸🇦  
2           🇪🇬  
3           🇺🇾  
4           🇵🇹  
5           🇪🇸  
6           🇲🇦  
7           🇮🇷  
8           🇫🇷  
9           🇦🇺  
10          🇵🇪  
11          🇩🇰  
12          🇦🇷  
13          🇮🇸  
14          🇭🇷  
15          🇳🇬  
16          🇧🇷  


In [10]:
groups

index     name winner runnerup  \
0     a  Group A      4        1   
1     b  Group B      6        5   
2     c  Group C      9       12   
3     d  Group D     15       13   
4     e  Group E     17       18   
5     f  Group F     23       22   
6     g  Group G     25       28   
7     h  Group H     31       32   

                                             matches  
0  [{'name': 1, 'type': 'group', 'home_team': 1, ...  
1  [{'name': 3, 'type': 'group', 'home_team': 5, ...  
2  [{'name': 5, 'type': 'group', 'home_team': 9, ...  
3  [{'name': 7, 'type': 'group', 'home_team': 13,...  
4  [{'name': 9, 'type': 'group', 'home_team': 17,...  
5  [{'name': 11, 'type': 'group', 'home_team': 21...  
6  [{'name': 13, 'type': 'group', 'home_team': 25...  
7  [{'name': 15, 'type': 'group', 'home_team': 29...

# 2018

In [11]:
groups = groups.rename({"name": "round"}, axis= 1)

In [12]:
groups_matches = pd.concat([pd.DataFrame(x) for x in groups['matches']], keys=groups['round']) \
       .reset_index(level=1, drop=True).reset_index()

In [13]:
knockout = knockout.rename({"name": "round"}, axis= 1)

In [14]:
knockout_matches = pd.concat([pd.DataFrame(x) for x in knockout['matches']], keys=knockout['round']) \
       .reset_index(level=1, drop=True).reset_index()

In [15]:
fifa_2018 = pd.concat([groups_matches, knockout_matches])

In [16]:
team_dict = dict(zip(teams.id, teams.name))
city_dict = dict(zip(stadiums.id, stadiums.city))

In [17]:
fifa_2018['home_team'] = fifa_2018['home_team'].map(team_dict)
fifa_2018['away_team'] = fifa_2018['away_team'].map(team_dict)
fifa_2018['stadium'] = fifa_2018['stadium'].map(city_dict)

In [18]:
fifa_2018 = fifa_2018.rename({"stadium": "city"}, axis= 1)

In [19]:
fifa_2018 = fifa_2018[["name", "home_team", "away_team", "home_result", "away_result", "date", "round", "city"]]

In [20]:
fifa_2018["edition"] = 2018

In [21]:
fifa_2018 = fifa_2018.rename({"name": "id_match"}, axis= 1)

In [22]:
fifa_2018

id_match home_team     away_team  home_result  away_result  \
0          1    Russia  Saudi Arabia            5            0   
1          2     Egypt       Uruguay            0            1   
2         17    Russia         Egypt            3            1   
3         18   Uruguay  Saudi Arabia            1            0   
4         33   Uruguay        Russia            3            0   
..       ...       ...           ...          ...          ...   
11        60    Sweden       England            0            2   
12        61    France       Belgium            1            0   
13        62   Croatia       England            2            1   
14        63   Belgium       England            2            0   
15        64    France       Croatia            4            2   

                         date                 round              city  edition  
0   2018-06-14T18:00:00+03:00               Group A            Moscow     2018  
1   2018-06-15T17:00:00+05:00               Group A     Yekaterinburg     2018  
2   2018-06-19T21:00:00+03:00               Group A  Saint Petersburg     2018  
3   2018-06-20T18:00:00+03:00               Group A     Rostov-on-Don     2018  
4   2018-06-25T18:00:00+04:00               Group A            Samara     2018  
..                        ...                   ...               ...      ...  
11  2018-07-07T18:00:00+04:00        Quarter-finals            Samara     2018  
12  2018-07-10T21:00:00+03:00           Semi-finals  Saint Petersburg     2018  
13  2018-07-11T21:00:00+03:00           Semi-finals            Moscow     2018  
14  2018-07-14T17:00:00+03:00  Third place play-off  Saint Petersburg     2018  
15  2018-07-15T18:00:00+03:00                 Final            Moscow     2018  

[64 rows x 9 columns]

# 2010

In [23]:
matches

edition        round      score                     team1  \
0     1930-URUGUAY  GROUP_STAGE  4-1 (3-0)                    France   
1     1930-URUGUAY  GROUP_STAGE  3-0 (2-0)                       USA   
2     1930-URUGUAY  GROUP_STAGE  2-1 (2-0)  Yugoslavia (Југославија)   
3     1930-URUGUAY  GROUP_STAGE  3-1 (1-0)         Romania (România)   
4     1930-URUGUAY  GROUP_STAGE  1-0 (0-0)                 Argentina   
...            ...          ...        ...                       ...   
7294   2014-BRAZIL    1/4_FINAL        (C)                         3   
7295   2014-BRAZIL    1/2_FINAL        (X)                         A   
7296   2014-BRAZIL    1/2_FINAL        (Y)                         C   
7297   2014-BRAZIL   PLACES_3&4        xxx                   LOSER X   
7298   2014-BRAZIL       _FINAL        xxx                  WINNER X   

                 team2                               url            venue  \
0      Mexico (México)     1930_URUGUAY_FS.htm#1-WC-30-I      Montevideo.   
1     Belgium (België)    1930_URUGUAY_FS.htm#13-WC-30-I      Montevideo.   
2      Brazil (Brasil)     1930_URUGUAY_FS.htm#7-WC-30-I      Montevideo.   
3          Peru (Perú)    1930_URUGUAY_FS.htm#10-WC-30-I      Montevideo.   
4               France     1930_URUGUAY_FS.htm#2-WC-30-I      Montevideo.   
...                ...                               ...              ...   
7294                 4  2014_BRAZIL_FS.htm#832-WC-14-III        Salvador.   
7295                 B   2014_BRAZIL_FS.htm#833-WC-14-IV  Belo Horizonte.   
7296                 D   2014_BRAZIL_FS.htm#834-WC-14-IV       São_Paulo.   
7297           LOSER Y    2014_BRAZIL_FS.htm#835-WC-14-V        Brasília.   
7298          WINNER Y   2014_BRAZIL_FS.htm#836-WC-14-VI  Rio de Janeiro.   

      year  
0     1930  
1     1930  
2     1930  
3     1930  
4     1930  
...    ...  
7294  2014  
7295  2014  
7296  2014  
7297  2014  
7298  2014  

[7299 rows x 8 columns]

In [24]:
matches = matches.replace(["[\(\[].*?[\)\]]", "xxx"], "",regex=True)

In [25]:
matches = matches.replace(r'^\s*$', np.nan, regex=True)

In [26]:
matches = matches.dropna()

In [27]:
matches['result'] = matches['score'].str.split('-')

In [28]:
truc = []
trac = []
for i in matches['result']:
    truc.append(i[0])
    trac.append(i[1])
matches['home_result'] = truc
matches['away_result'] = trac

In [29]:
matches = matches.reset_index()

In [30]:
matches = matches.rename({"index": "id_match", \
                          "team1": "home_team", \
                          "team2": "away_team", \
                          "year": "date", \
                          "venue": "city"} \
                         , axis= 1)

In [31]:
matches['edition'] = matches['date']

In [32]:
matches['city'] = matches['city'].str.replace('.', '')

In [33]:
fifa_2010 = matches[["id_match", "home_team", "away_team", "home_result", "away_result", "date", "round", "city", "edition"]]

In [34]:
fifa_2010 = fifa_2010[~fifa_2010['edition'].isin([2011, 2012, 2013, 2014])]

In [35]:
fifa_2010

id_match     home_team     away_team home_result away_result  date  \
0            0        France       Mexico            4          1   1930   
1            1           USA      Belgium            3          0   1930   
2            2   Yugoslavia        Brazil            2          1   1930   
3            3      Romania          Peru            3          1   1930   
4            4     Argentina        France           1          0   1930   
...        ...           ...           ...         ...         ...   ...   
6410      6410      Paraguay        Spain            0          1   2010   
6411      6411       Uruguay  Netherlands            2          3   2010   
6412      6412      Germany         Spain            0          1   2010   
6413      6413       Uruguay      Germany            2          3   2010   
6414      6414  Netherlands         Spain            0         1    2010   

            round            city  edition  
0     GROUP_STAGE      Montevideo     1930  
1     GROUP_STAGE      Montevideo     1930  
2     GROUP_STAGE      Montevideo     1930  
3     GROUP_STAGE      Montevideo     1930  
4     GROUP_STAGE      Montevideo     1930  
...           ...             ...      ...  
6410    1/4_FINAL    Johannesburg     2010  
6411    1/2_FINAL       Cape Town     2010  
6412    1/2_FINAL          Durban     2010  
6413   PLACES_3&4  Port_Elizabeth     2010  
6414       _FINAL    Johannesburg     2010  

[6415 rows x 9 columns]

# 2014

In [36]:
wolrdcup

Year              Datetime                     Stage              Stadium  \
0   2014  12 Jun 2014 - 17:00                    Group A   Arena de Sao Paulo   
1   2014  13 Jun 2014 - 13:00                    Group A    Estadio das Dunas   
2   2014  13 Jun 2014 - 16:00                    Group B     Arena Fonte Nova   
3   2014  13 Jun 2014 - 18:00                    Group B       Arena Pantanal   
4   2014  14 Jun 2014 - 13:00                    Group C     Estadio Mineirao   
..   ...                   ...                       ...                  ...   
75  2014  05 Jul 2014 - 17:00             Quarter-finals     Arena Fonte Nova   
76  2014  08 Jul 2014 - 17:00                Semi-finals     Estadio Mineirao   
77  2014  09 Jul 2014 - 17:00                Semi-finals   Arena de Sao Paulo   
78  2014  12 Jul 2014 - 17:00   Play-off for third place     Estadio Nacional   
79  2014  13 Jul 2014 - 16:00                      Final  Estadio do Maracana   

               City Home Team Name  Home Team Goals  Away Team Goals  \
0        Sao Paulo          Brazil                3                1   
1            Natal          Mexico                1                0   
2         Salvador           Spain                1                5   
3           Cuiaba           Chile                3                1   
4   Belo Horizonte        Colombia                3                0   
..              ...            ...              ...              ...   
75        Salvador     Netherlands                0                0   
76  Belo Horizonte          Brazil                1                7   
77       Sao Paulo     Netherlands                0                0   
78        Brasilia          Brazil                0                3   
79  Rio De Janeiro         Germany                1                0   

   Away Team Name                         Win conditions  Attendance  \
0         Croatia                                            62103.0   
1        Cameroon                                            39216.0   
2     Netherlands                                            48173.0   
3       Australia                                            40275.0   
4          Greece                                            57174.0   
..            ...                                    ...         ...   
75     Costa Rica  Netherlands win on penalties (4 - 3)      51179.0   
76        Germany                                            58141.0   
77      Argentina    Argentina win on penalties (2 - 4)      63267.0   
78    Netherlands                                            68034.0   
79      Argentina          Germany win after extra time      74738.0   

    Half-time Home Goals  Half-time Away Goals                 Referee  \
0                      1                     1  NISHIMURA Yuichi (JPN)   
1                      0                     0     ROLDAN Wilmar (COL)   
2                      1                     1    Nicola RIZZOLI (ITA)   
3                      2                     1   Noumandiez DOUE (CIV)   
4                      1                     0       GEIGER Mark (USA)   
..                   ...                   ...                     ...   
75                     0                     0   Ravshan IRMATOV (UZB)   
76                     0                     5   RODRIGUEZ Marco (MEX)   
77                     0                     0      C�neyt �AKIR (TUR)   
78                     0                     2   HAIMOUDI Djamel (ALG)   
79                     0                     0    Nicola RIZZOLI (ITA)   

                    Assistant 1                    Assistant 2  RoundID  \
0             SAGARA Toru (JPN)           NAGI Toshiyuki (JPN)   255931   
1        CLAVIJO Humberto (COL)             DIAZ Eduardo (COL)   255931   
2         Renato FAVERANI (ITA)           Andrea STEFANI (ITA)   255931   
3          YEO Songuifolo (CIV)  BIRUMUSHAHU Jean Claude (BDI)   255931   
4               HURD Sean (USA)      

In [37]:
fifa_2014 = wolrdcup.rename({"MatchID": "id_match", \
                          "Home Team Name": "home_team", \
                          "Away Team Name": "away_team", \
                          "Home Team Goals": "home_result", \
                          "Away Team Goals": "away_result", \
                          "Datetime": "date", \
                          "Stage": "round", \
                          "City": "city", \
                          "Year": "edition" \
                            } \
                         , axis= 1)

In [38]:
fifa_2014 = fifa_2014[["id_match", "home_team", "away_team", "home_result", "away_result", "date", "round", "city", "edition"]]

In [39]:
fifa_2014

id_match    home_team    away_team  home_result  away_result  \
0   300186456       Brazil      Croatia            3            1   
1   300186492       Mexico     Cameroon            1            0   
2   300186510        Spain  Netherlands            1            5   
3   300186473        Chile    Australia            3            1   
4   300186471     Colombia       Greece            3            0   
..        ...          ...          ...          ...          ...   
75  300186488  Netherlands   Costa Rica            0            0   
76  300186474       Brazil      Germany            1            7   
77  300186490  Netherlands    Argentina            0            0   
78  300186502       Brazil  Netherlands            0            3   
79  300186501      Germany    Argentina            1            0   

                    date                     round             city  edition  
0   12 Jun 2014 - 17:00                    Group A       Sao Paulo      2014  
1   13 Jun 2014 - 13:00                    Group A           Natal      2014  
2   13 Jun 2014 - 16:00                    Group B        Salvador      2014  
3   13 Jun 2014 - 18:00                    Group B          Cuiaba      2014  
4   14 Jun 2014 - 13:00                    Group C  Belo Horizonte      2014  
..                   ...                       ...              ...      ...  
75  05 Jul 2014 - 17:00             Quarter-finals        Salvador      2014  
76  08 Jul 2014 - 17:00                Semi-finals  Belo Horizonte      2014  
77  09 Jul 2014 - 17:00                Semi-finals       Sao Paulo      2014  
78  12 Jul 2014 - 17:00   Play-off for third place        Brasilia      2014  
79  13 Jul 2014 - 16:00                      Final  Rio De Janeiro      2014  

[80 rows x 9 columns]

# 2022

In [40]:
fifa

team1         team2 possession team1 possession team2  \
0           QATAR       ECUADOR              42%              50%   
1         ENGLAND          IRAN              72%              19%   
2         SENEGAL   NETHERLANDS              44%              45%   
3   UNITED STATES         WALES              51%              39%   
4       ARGENTINA  SAUDI ARABIA              64%              24%   
..            ...           ...              ...              ...   
59        ENGLAND        FRANCE              54%              36%   
60      ARGENTINA       CROATIA              34%              54%   
61         FRANCE       MOROCCO              34%              55%   
62        CROATIA       MOROCCO              45%              45%   
63      ARGENTINA        FRANCE              46%              40%   

   possession in contest  number of goals team1  number of goals team2  \
0                     8%                      0                      2   
1                     9%                      6                      2   
2                    11%                      0                      2   
3                    10%                      1                      1   
4                    12%                      1                      2   
..                   ...                    ...                    ...   
59                   10%                      1                      2   
60                   12%                      3                      0   
61                   11%                      2                      0   
62                   10%                      2                      1   
63                   14%                      3                      3   

           date     hour                  category  ...  \
0   20 NOV 2022  17 : 00                   Group A  ...   
1   21 NOV 2022  14 : 00                   Group B  ...   
2   21 NOV 2022  17 : 00                   Group A  ...   
3   21 NOV 2022  20 : 00                   Group B  ...   
4   22 NOV 2022  11 : 00                   Group C  ...   
..          ...      ...                       ...  ...   
59  10 DEC 2022  20 : 00             Quarter-final  ...   
60  13 DEC 2022  20 : 00                Semi-final  ...   
61  14 DEC 2022  20 : 00                Semi-final  ...   
62  17 DEC 2022  16 : 00  Play-off for third place  ...   
63  18 DEC 2022  16 : 00                     Final  ...   

    penalties scored team1  penalties scored team2  goal preventions team1  \
0                        0                       1                       6   
1                        0                       1                       8   
2                        0                       0                       9   
3                        0                       1                       7   
4                        1                       0                       4   
..                     ...                     ...                     ...   
59                       1                       0                       9   
60                       1                       0                      12   
61                       0                       0                      13   
62                       0                       0                       7   
63                       1                       2                      11   

    goal preventions team2  own goals team1  own goals team2  \
0                        5                0                0   
1                       13                0                0   
2                       15                0                0   
3                        7                0                0   
4                       14                0                0   
..                     ...              ...              ...   
59                      15                0                0   
60                      10                0                0   
61                      14                0                0   
62      

In [41]:
fife

match dayofweek           match_time      home_team     away_team  \
0       1       Sun  2022-11-20 19:00:00          Qatar       Ecuador   
1       2       Mon  2022-11-21 16:00:00        England       IR Iran   
2       3       Mon  2022-11-21 19:00:00        Senegal   Netherlands   
3       4       Mon  2022-11-21 22:00:00  United States         Wales   
4       5       Tue  2022-11-22 13:00:00      Argentina  Saudi Arabia   
..    ...       ...                  ...            ...           ...   
59     60       Sat  2022-12-10 22:00:00        England        France   
60     61       Tue  2022-12-13 22:00:00      Argentina       Croatia   
61     62       Wed  2022-12-14 22:00:00         France       Morocco   
62     63       Sat  2022-12-17 18:00:00        Croatia       Morocco   
63     64       Sun  2022-12-18 18:00:00      Argentina        France   

    home_xg  away_xg        score  attendance                          venue  \
0       0.3      1.2          0–2       67372                Al Bayt Stadium   
1       2.1      1.4          6–2       45334  Khalifa International Stadium   
2       0.9      0.7          0–2       41721             Al Thumama Stadium   
3       0.8      1.5          1–1       43418          Ahmed bin Ali Stadium   
4       2.2      0.1          1–2       88012          Lusail Iconic Stadium   
..      ...      ...          ...         ...                            ...   
59      2.4      0.9          1–2       68895                Al Bayt Stadium   
60      2.3      0.5          3–0       88966          Lusail Iconic Stadium   
61      1.9      0.9          2–0       68294                Al Bayt Stadium   
62      0.7      1.2          2–1       44137  Khalifa International Stadium   
63      3.3      2.2  (4) 3–3 (2)       88966          Lusail Iconic Stadium   

    ... home_clearances away_clearances home_offsides away_offsides home_gks  \
0   ...              18               7             3             4        4   
1   ...               4              23             2             2        5   
2   ...              32              16             2             1        6   
3   ...              18              31             1             1        5   
4   ...              11              26            10             1        3   
..  ...             ...             ...           ...           ...      ...   
59  ...              13              16             1             2        9   
60  ...              22               6             1             0        9   
61  ...              16               3             4             3        8   
62  ...              19              16             2             2       12   
63  ...              24              18             4             4        6   

   away_gks home_throw_ins  away_throw_ins  home_long_balls  away_long_balls  
0         7             20              17               51               70  
1         5             18              17               87               50  
2        10             17              28               64               66  
3        10             22              25               56               90  
4         9             24              13               55               59  
..      ...            ...             ...              ...              ...  
59        5             16              14               61               62  
60        4             19              21               49               43  
61       10             14              18               55               40  
62       11             14              19               53               46  
63       13             26              21               73               94  

[64 rows x 53 columns]

In [42]:
fife['round'] = fifa['category']
fife['date'] = 2022
fife['edition'] = 2022

In [43]:
fife['result'] = fife['score'].str.split('–')

In [44]:
troc = []
tric = []
for i in fife['result']:
    troc.append(i[0])
    tric.append(i[1])
fife['home_result'] = troc
fife['away_result'] = tric

In [45]:
fife

match dayofweek           match_time      home_team     away_team  \
0       1       Sun  2022-11-20 19:00:00          Qatar       Ecuador   
1       2       Mon  2022-11-21 16:00:00        England       IR Iran   
2       3       Mon  2022-11-21 19:00:00        Senegal   Netherlands   
3       4       Mon  2022-11-21 22:00:00  United States         Wales   
4       5       Tue  2022-11-22 13:00:00      Argentina  Saudi Arabia   
..    ...       ...                  ...            ...           ...   
59     60       Sat  2022-12-10 22:00:00        England        France   
60     61       Tue  2022-12-13 22:00:00      Argentina       Croatia   
61     62       Wed  2022-12-14 22:00:00         France       Morocco   
62     63       Sat  2022-12-17 18:00:00        Croatia       Morocco   
63     64       Sun  2022-12-18 18:00:00      Argentina        France   

    home_xg  away_xg        score  attendance                          venue  \
0       0.3      1.2          0–2       67372                Al Bayt Stadium   
1       2.1      1.4          6–2       45334  Khalifa International Stadium   
2       0.9      0.7          0–2       41721             Al Thumama Stadium   
3       0.8      1.5          1–1       43418          Ahmed bin Ali Stadium   
4       2.2      0.1          1–2       88012          Lusail Iconic Stadium   
..      ...      ...          ...         ...                            ...   
59      2.4      0.9          1–2       68895                Al Bayt Stadium   
60      2.3      0.5          3–0       88966          Lusail Iconic Stadium   
61      1.9      0.9          2–0       68294                Al Bayt Stadium   
62      0.7      1.2          2–1       44137  Khalifa International Stadium   
63      3.3      2.2  (4) 3–3 (2)       88966          Lusail Iconic Stadium   

    ... home_throw_ins away_throw_ins home_long_balls away_long_balls  \
0   ...             20             17              51              70   
1   ...             18             17              87              50   
2   ...             17             28              64              66   
3   ...             22             25              56              90   
4   ...             24             13              55              59   
..  ...            ...            ...             ...             ...   
59  ...             16             14              61              62   
60  ...             19             21              49              43   
61  ...             14             18              55              40   
62  ...             14             19              53              46   
63  ...             26             21              73              94   

                       round  date edition          result  home_result  \
0                    Group A  2022    2022          [0, 2]            0   
1                    Group B  2022    2022          [6, 2]            6   
2                    Group A  2022    2022          [0, 2]            0   
3                    Group B  2022    2022          [1, 1]            1   
4                    Group C  2022    2022          [1, 2]            1   
..                       ...   ...     ...             ...          ...   
59             Quarter-final  2022    2022          [1, 2]            1   
60                Semi-final  2022    2022          [3, 0]            3   
61                Semi-final  2022    2022          [2, 0]            2   
62  Play-off for third place  2022    2022          [2, 1]            2   
63                     Final  2022    2022  [(4) 3, 3 (2)]        (4) 3   

    away_result  
0             2  
1             2  
2             2  
3             1  
4             2  
..          ...  
59            2  
60            0  
61            0  
62            1  
63        3 (2)  

[64 rows x 59 columns]

In [46]:
fifa_2022 = fife.rename({"match": "id_match", \
                          "Stage": "round", \
                          "venue": "city", \
                            } \
                         , axis= 1)

In [47]:
fifa_2022 = fifa_2022[["id_match", "home_team", "away_team", "home_result", "away_result", "date", "round", "city", "edition"]]

In [48]:
fifa_2022

id_match      home_team     away_team home_result away_result  date  \
0          1          Qatar       Ecuador           0           2  2022   
1          2        England       IR Iran           6           2  2022   
2          3        Senegal   Netherlands           0           2  2022   
3          4  United States         Wales           1           1  2022   
4          5      Argentina  Saudi Arabia           1           2  2022   
..       ...            ...           ...         ...         ...   ...   
59        60        England        France           1           2  2022   
60        61      Argentina       Croatia           3           0  2022   
61        62         France       Morocco           2           0  2022   
62        63        Croatia       Morocco           2           1  2022   
63        64      Argentina        France       (4) 3       3 (2)  2022   

                       round                           city  edition  
0                    Group A                Al Bayt Stadium     2022  
1                    Group B  Khalifa International Stadium     2022  
2                    Group A             Al Thumama Stadium     2022  
3                    Group B          Ahmed bin Ali Stadium     2022  
4                    Group C          Lusail Iconic Stadium     2022  
..                       ...                            ...      ...  
59             Quarter-final                Al Bayt Stadium     2022  
60                Semi-final          Lusail Iconic Stadium     2022  
61                Semi-final                Al Bayt Stadium     2022  
62  Play-off for third place  Khalifa International Stadium     2022  
63                     Final          Lusail Iconic Stadium     2022  

[64 rows x 9 columns]

# EL COPA DEL MUNDOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO AYAYAYAYAYAYAAYYAYAY AQUI AQUI

In [49]:
fifa_2014['date'] = fifa_2014['edition']

In [50]:
fifa_2018['date'] = fifa_2018['edition']

In [51]:
fifa_2018['id_match'] = fifa_2018['id_match'] + fifa_2014['id_match'].max()

In [52]:
fifa_2022['id_match'] = fifa_2022['id_match'] + fifa_2018['id_match'].max()

In [53]:
EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO = pd.concat([fifa_2010, fifa_2014, fifa_2018, fifa_2022])

In [54]:
EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO = EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO.replace("[\(\[].*?[\)\]]", "",regex=True)

In [55]:
EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO['home_team'] = EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO['home_team'].str.strip()
EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO['away_team'] = EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO['away_team'].str.strip()

In [56]:
EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO

id_match   home_team away_team home_result away_result  date  \
0           0      France    Mexico           4          1   1930   
1           1         USA   Belgium           3          0   1930   
2           2  Yugoslavia    Brazil           2          1   1930   
3           3     Romania      Peru           3          1   1930   
4           4   Argentina    France           1          0   1930   
..        ...         ...       ...         ...         ...   ...   
59  300186639     England    France           1           2  2022   
60  300186640   Argentina   Croatia           3           0  2022   
61  300186641      France   Morocco           2           0  2022   
62  300186642     Croatia   Morocco           2           1  2022   
63  300186643   Argentina    France           3          3   2022   

                       round                           city  edition  
0                GROUP_STAGE                     Montevideo     1930  
1                GROUP_STAGE                     Montevideo     1930  
2                GROUP_STAGE                     Montevideo     1930  
3                GROUP_STAGE                     Montevideo     1930  
4                GROUP_STAGE                     Montevideo     1930  
..                       ...                            ...      ...  
59             Quarter-final                Al Bayt Stadium     2022  
60                Semi-final          Lusail Iconic Stadium     2022  
61                Semi-final                Al Bayt Stadium     2022  
62  Play-off for third place  Khalifa International Stadium     2022  
63                     Final          Lusail Iconic Stadium     2022  

[6623 rows x 9 columns]

# kestions

## quel pays a gagné le + de matchs en coupe du monde ?

In [57]:
bip = EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO.copy()

In [58]:
bip['home_result'] = bip['home_result'].astype(int)
bip['away_result'] = bip['away_result'].astype(int)

In [59]:
bip

id_match   home_team away_team  home_result  away_result  date  \
0           0      France    Mexico            4            1  1930   
1           1         USA   Belgium            3            0  1930   
2           2  Yugoslavia    Brazil            2            1  1930   
3           3     Romania      Peru            3            1  1930   
4           4   Argentina    France            1            0  1930   
..        ...         ...       ...          ...          ...   ...   
59  300186639     England    France            1            2  2022   
60  300186640   Argentina   Croatia            3            0  2022   
61  300186641      France   Morocco            2            0  2022   
62  300186642     Croatia   Morocco            2            1  2022   
63  300186643   Argentina    France            3            3  2022   

                       round                           city  edition  
0                GROUP_STAGE                     Montevideo     1930  
1                GROUP_STAGE                     Montevideo     1930  
2                GROUP_STAGE                     Montevideo     1930  
3                GROUP_STAGE                     Montevideo     1930  
4                GROUP_STAGE                     Montevideo     1930  
..                       ...                            ...      ...  
59             Quarter-final                Al Bayt Stadium     2022  
60                Semi-final          Lusail Iconic Stadium     2022  
61                Semi-final                Al Bayt Stadium     2022  
62  Play-off for third place  Khalifa International Stadium     2022  
63                     Final          Lusail Iconic Stadium     2022  

[6623 rows x 9 columns]

In [60]:
conditions = [
    bip['home_result'] == bip['away_result'],
    bip['home_result'] > bip['away_result'],
    bip['home_result'] < bip['away_result']
]

categories = ['draw', bip['home_team'], bip['away_team']]

# Use np.select() to create the new column
bip['VOCTOIRY'] = np.select(conditions, categories, default='Unknown')

In [61]:
bip['VOCTOIRY'].value_counts()

VOCTOIRY
draw              1408
Brazil             133
Mexico             109
Argentina          108
Italy              106
                  ... 
Laos                 1
Guinea-Bissau        1
Saarland             1
Western Samoa        1
Korea Republic       1
Name: count, Length: 203, dtype: int64

c'est le brésil qui a le plus de match gagnés avec 123 victoires

## quel est le top 5 des vainqueurs de coupe du monde (finale) ?

In [62]:
bop = EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO.copy()

In [63]:
bop['round'] = bop['round'].str.replace('_', '')
# bop['round'] = bop['round'].str.replace('FIRST', 'final')
# bop['round'] = bop['round'].str.replace('_', '')
bop['round'] = bop['round'].str.replace('ROUND', '').str.lower()

In [64]:
bop['round'].value_counts()

round
preliminary-europe          2064
preliminary-africa           980
preliminary-asia             828
preliminary-n/c.america      823
preliminary-s.america        557
groupstage                   532
preliminary-as./o.           173
preliminary-oceania          135
1/4final                      58
1/8final                      56
round of 16                   32
1/2final                      32
final                         28
semifinalstage                24
preliminary-o./as.            22
group c                       18
group b                       18
group h                       18
group g                       18
group f                       18
group e                       18
group d                       18
group a                       18
places3&4                     17
first                         17
preliminary-euro/as.          16
preliminary-eur./n.e.         15
preliminary-eu./afr.          13
quarterfinalstage             12
quarter-finals                12
prel

In [65]:
bop = bop[bop['round'].isin(['final'])]

In [66]:
bop = bop.drop_duplicates()

In [67]:
bop['home_result'] = bop['home_result'].astype(int)
bop['away_result'] = bop['away_result'].astype(int)

In [68]:
conditions = [
    bop['home_result'] > bop['away_result'],
    bop['home_result'] < bop['away_result'],
    bop['home_result'] == bop['away_result']
]

categories = [bop['home_team'], bop['away_team'], 'draw']

# Use np.select() to create the new column
bop['VOCTOIRY'] = np.select(conditions, categories, default='Unknown')

In [69]:
bop['VOCTOIRY'].value_counts()

VOCTOIRY
Brazil       6
draw         4
Uruguay      3
Italy        3
FRG          3
Argentina    2
France       2
Sweden       1
England      1
Spain        1
Germany      1
Name: count, dtype: int64

le top 5 est dans l'ordre respectif avec leur scores\
Brazil     6 \
Uruguay      3 \
Italy        3 \
FRG          3 \
Argentina    2 

## quel est le plus grand écart de buts dans un match ?

In [70]:
bep = EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO.copy()

In [71]:
bep['away_result'] = bep['away_result'].astype(int)
bep['home_result'] = bep['home_result'].astype(int)

In [72]:
bep['ecart'] = bep['home_result'] - bep['away_result']

In [73]:
bep['ecart'] = bep['ecart'].astype(str).str.replace('-', '')

In [74]:
bep['ecart'] = bep['ecart'].astype(int)

In [75]:
bep.sort_values(by='ecart', ascending=False).head(1)

id_match  home_team       away_team  home_result  away_result  date  \
4160      4160  Australia  American Samoa           31            0  2001   

                    round           city  edition  ecart  
4160  PRELIMINARY-Oceania  Coffs Harbour     2001     31

Le plus grand écart est de 31 point avec l'australie contre l'amerique somoayaine

## combien de matchs nuls au total ?

Il y a 1408 matchs nuls

## quel pays est allé le plus souvent en phase finale (à partir des 8ème de finale) de cdm ?

In [76]:
bap = EL_COPA_DEL_MUNDOOOOOOOOOOOOOOOOOOOOOOOOO.copy()

In [77]:
search = ['preliminary', 'group', 'first']

In [78]:
bap = bap[~bap['round'].str.contains('|'.join(search))]

In [79]:
bap['round'].value_counts()

round
PRELIMINARY-Europe          2064
PRELIMINARY-Africa           980
PRELIMINARY-Asia             828
PRELIMINARY-N/C.America      823
PRELIMINARY-S.America        557
GROUP_STAGE                  532
PRELIMINARY-As./O.           173
PRELIMINARY-Oceania          135
1/4_FINAL                     58
1/8_FINAL                     56
Round of 16                   32
1/2_FINAL                     32
SEMIFINAL_STAGE               24
PRELIMINARY-O./As.            22
Group C                       18
Group D                       18
_FINAL                        18
Group H                       18
Group G                       18
Group F                       18
Group E                       18
Group A                       18
Group B                       18
PLACES_3&4                    17
FIRST                         17
PRELIMINARY-Euro/As.          16
PRELIMINARY-Eur./N.E.         15
PRELIMINARY-Eu./Afr.          13
QUARTERFINAL_STAGE            12
Quarter-finals                12
PREL

In [80]:
trec = pd.concat([bap['home_team'], bap['away_team']])

In [81]:
trec.value_counts()

Brazil               210
Mexico               202
Argentina            194
Uruguay              178
Italy                170
                    ... 
East Timor             2
Dahomey                2
Guam                   2
Dutch East Indies      1
Guayana                1
Name: count, Length: 234, dtype: int64

Le brésil est arrivé le plus souvent en phase de finale avec 46 occurences